lib/ui/window/platform_configuration_unittests.cc

```diff
@@ -332,5 +488,68 @@ TEST_F(PlatformConfigurationTest, SetDartPerformanceMode) {
   DestroyShell(std::move(shell), task_runners);
 }
 
+TEST_F(PlatformConfigurationTest, OutOfScopeRenderCallsAreIgnored) {
+  Settings settings = CreateSettingsForFixture();
+  TaskRunners task_runners = GetTaskRunnersForFixture();
+
+  MockRuntimeDelegate client;
+  auto platform_message_handler =
+      std::make_shared<MockPlatformMessageHandler>();
+  EXPECT_CALL(client, GetPlatformMessageHandler())
+      .WillOnce(Return(platform_message_handler));
+  // Render should not be called.
+  EXPECT_CALL(client, Render(_, _)).Times(Exactly(0));
+
+  auto runtime_controller_context =
+      RuntimeControllerContext::Create(settings, task_runners, client);
+
+  auto configuration = RunConfiguration::InferFromSettings(settings);
+  configuration.SetEntrypoint("incorrectImmediateRender");
+  runtime_controller_context->LaunchRootIsolate(configuration, [&] {
+    runtime_controller_context->Controller().AddView(
+        kImplicitViewId, ViewportMetrics(
+                             /*pixel_ratio=*/1.0, /*width=*/20, /*height=*/20,
+                             /*touch_slop=*/2, /*display_id=*/0));
+  });
+
+  // Wait for the Dart main function to end.
+  fml::AutoResetWaitableEvent latch;
+  PostSync(task_runners.GetUITaskRunner(), [&]() { latch.Signal(); });
+  latch.Wait();
+}
+
+TEST_F(PlatformConfigurationTest, DuplicateRenderCallsAreIgnored) {
+  Settings settings = CreateSettingsForFixture();
+  TaskRunners task_runners = GetTaskRunnersForFixture();
+
+  MockRuntimeDelegate client;
+  auto platform_message_handler =
+      std::make_shared<MockPlatformMessageHandler>();
+  EXPECT_CALL(client, GetPlatformMessageHandler())
+      .WillOnce(Return(platform_message_handler));
+  // Render should only be called once, because the second call is ignored.
+  EXPECT_CALL(client, Render(_, _)).Times(Exactly(1));
+
+  auto runtime_controller_context =
+      RuntimeControllerContext::Create(settings, task_runners, client);
+
+  auto configuration = RunConfiguration::InferFromSettings(settings);
+  configuration.SetEntrypoint("incorrectDoubleRender");
+  runtime_controller_context->LaunchRootIsolate(configuration, [&] {
+    runtime_controller_context->Controller().AddView(
+        kImplicitViewId, ViewportMetrics(
+                             /*pixel_ratio=*/1.0, /*width=*/20, /*height=*/20,
+                             /*touch_slop=*/2, /*display_id=*/0));
+  });
+
+  // Wait for the Dart main function to end, which means the callbacks have been
+  // registered.
+  fml::AutoResetWaitableEvent latch;
+  PostSync(task_runners.GetUITaskRunner(), [&]() { latch.Signal(); });
+  latch.Wait();
+
+  runtime_controller_context->Controller().BeginFrame(fml::TimePoint::Now(), 0);
```

Comment:
Should this test wait until `onDrawFrame` exits? Perhaps `onDrawFrame` should call `_finish` and this should wait for that?

Reply:
On second thought, that's not necessary. `BeginFrame` calls the callbacks synchronously. Could you add a quick comment that this will call `onBeginFrame` and `onDrawFrame` synchronously?